kotlinx-coroutines-core/jvm/src/Dispatchers.kt

```diff
@@ -106,5 +107,32 @@ public actual object Dispatchers {
      * typically execution continues in the same thread.
      */
     @JvmStatic
-    public val IO: CoroutineDispatcher = DefaultScheduler.IO
+    public val IO: CoroutineDispatcher = createIoDispatcher()
 }
+
+// Dispatchers initialization
+
+internal const val COROUTINES_SCHEDULER_PROPERTY_NAME = "kotlinx.coroutines.scheduler"
+
+// @JvmField: Don't use JvmField here to enable R8 optimizations via "assumenosideeffects"
+internal val DEFAULT_SCHEDULER = systemProp(COROUTINES_SCHEDULER_PROPERTY_NAME).let { value ->
+    when (value) {
+        null, "", "on" -> true
+        "off" -> false
+        else -> error("System property '$COROUTINES_SCHEDULER_PROPERTY_NAME' has unrecognized value '$value'")
+    }
+}
+
+// R8 optimization hook, not const on purpose
+// @JvmField: Don't use JvmField here to enable R8 optimizations via "assumenosideeffects"
+@Suppress("MayBeConstant")
+internal val IO_SCHEDULER = true
+
+@Suppress("NOTHING_TO_INLINE") // For better constant propagation in R8
+internal actual inline fun createDefaultDispatcher(): CoroutineDispatcher =
+    if (DEFAULT_SCHEDULER) DefaultScheduler else CommonPool
+
+@Suppress("NOTHING_TO_INLINE", "ConstantConditionIf") // For better constant propagation in R8
+internal inline fun createIoDispatcher(): CoroutineDispatcher =
+    if (IO_SCHEDULER) DefaultScheduler.IO else CommonPool
```

Comment:
`CommonPool` cannot be used as IO dispatcher.
Either separate FJP/executor (NB: daemon threads) should be created or (if we are willing to implement "pool view" API) `ManagedBlocker`-based dispatcher should be used

Reply:
The importance of a proper IO dispatcher for Android is questionable, but, indeed, it would be better to figure out a simple way to implement a better IO dispatcher. It might be based on some cached thread pool executor, for example. It will forfeit thread sharing, but that should not be an issue for Android anyway.

Reply:
>but that should not be an issue for Android anyway.

Apparently, it is an issue, according to different feedback.
We should try to minimize the number of threads, especially for "view" API.
I will investigate it further.

Reply:
Let's postone working of this PR until after 1.3.0 is out. We'll need to get a number of R8 fixes, too, in order to be able to trim the size further anyway. If I have time I'll see if I can separate some less questionable pieces of this big PR as smaller PRs.